zenoh/src/api/session.rs

```diff
@@ -244,22 +266,69 @@ impl SessionState {
             self.remote_key_to_expr(key_expr)
         }
     }
+
+    pub(crate) fn subscribers(&self, kind: SubscriberKind) -> &HashMap<Id, Arc<SubscriberState>> {
+        match kind {
+            SubscriberKind::Subscriber => &self.subscribers,
+            SubscriberKind::LivelinessSubscriber => &self.liveliness_subscribers,
+        }
+    }
+
+    pub(crate) fn subscribers_mut(
+        &mut self,
+        kind: SubscriberKind,
+    ) -> &mut HashMap<Id, Arc<SubscriberState>> {
+        match kind {
+            SubscriberKind::Subscriber => &mut self.subscribers,
+            SubscriberKind::LivelinessSubscriber => &mut self.liveliness_subscribers,
+        }
+    }
 }
 
 impl fmt::Debug for SessionState {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         write!(
             f,
-            "SessionState{{ subscribers: {} }}",
-            self.subscribers.len()
+            "SessionState{{ subscribers: {}, liveliness_subscribers: {} }}",
+            self.subscribers.len(),
+            self.liveliness_subscribers.len()
         )
     }
 }
 
 pub(crate) struct ResourceNode {
     pub(crate) key_expr: OwnedKeyExpr,
     pub(crate) subscribers: Vec<Arc<SubscriberState>>,
+    pub(crate) liveliness_subscribers: Vec<Arc<SubscriberState>>,
 }
+
+impl ResourceNode {
+    pub(crate) fn new(key_expr: OwnedKeyExpr) -> Self {
+        Self {
+            key_expr,
+            subscribers: Vec::new(),
+            liveliness_subscribers: Vec::new(),
+        }
+    }
+
+    pub(crate) fn subscribers(&self, kind: SubscriberKind) -> &Vec<Arc<SubscriberState>> {
```

Comment:
Could the return type be `&[Arc<SubscriberState>]`?

Reply:
As I believe we need to keep `subscribers_mut()` returning an `&mut Vec<_>` then I'd rather keep `subscribers()`as it is for homogeneity.